src/common/selftest/selftest_firstlayer.cpp

```diff
@@ -0,0 +1,355 @@
+static constexpr float nozzle_to_probe[] = NOZZLE_TO_PROBE_OFFSET;
+static constexpr float z_offset_def = nozzle_to_probe[AxisEnum::Z_AXIS];
+
+LoopResult CSelftestPart_FirstLayer::stateInitialDistanceInit() {
+    float diff = probe_offset.z - z_offset_def;
+    if ((diff <= -z_offset_step) || (diff >= z_offset_step)) {
+        IPartHandler::SetFsmPhase(PhasesSelftest::FirstLayer_use_val);
+        current_offset_is_default = false;
+        rResult.current_offset = probe_offset.z;
+    } else {
+        current_offset_is_default = true;
+    }
+
+    return LoopResult::RunNext;
+}
+
+LoopResult CSelftestPart_FirstLayer::stateInitialDistance() {
+    if (current_offset_is_default)
+        return LoopResult::RunNext;
+
+    switch (rStateMachine.GetButtonPressed()) {
+    case Response::No:
+        probe_offset.z = z_offset_def;
+        // don't return / break
+    case Response::Yes:
+        return LoopResult::RunNext;
+    default:
+        break;
+    }
+    return LoopResult::RunCurrent;
+}
+
+LoopResult CSelftestPart_FirstLayer::stateShowStartPrint() {
+    IPartHandler::SetFsmPhase(PhasesSelftest::FirstLayer_start_print);
+    return LoopResult::RunNext;
+}
+
+LoopResult CSelftestPart_FirstLayer::statePrintInit() {
+    // reset progress
+    set_var_sd_percent_done(0);
+
+    IPartHandler::SetFsmPhase(PhasesSelftest::FirstLayer_mbl);
+    const int temp_nozzle = Filaments::Current().nozzle;
+    temp_nozzle_preheat = Filaments::Current().nozzle_preheat;
+    temp_bed = Filaments::Current().heatbed;
+
+    // nozzle temperature preheat
+    thermalManager.setTargetHotend(temp_nozzle_preheat, 0);
+    marlin_server_set_temp_to_display(temp_nozzle);
+
+    // bed temperature
+    thermalManager.setTargetBed(temp_bed);
+    return LoopResult::RunNext;
+}
+
+LoopResult CSelftestPart_FirstLayer::stateWaitNozzle() {
+    std::array<char, sizeof("M109 R170")> gcode_buff; //safe to be local variable, queue.enqueue_one will copy it
+    snprintf(gcode_buff.begin(), gcode_buff.size(), "M109 R%d", temp_nozzle_preheat);
+    return queue.enqueue_one(gcode_buff.begin()) ? LoopResult::RunNext : LoopResult::RunCurrent;
+}
+
+LoopResult CSelftestPart_FirstLayer::stateWaitBed() {
+    std::array<char, sizeof("M190 S100")> gcode_buff; //safe to be local variable, queue.enqueue_one will copy
+    snprintf(gcode_buff.begin(), gcode_buff.size(), "M190 S%d", temp_bed);
+    return queue.enqueue_one(gcode_buff.begin()) ? LoopResult::RunNext : LoopResult::RunCurrent;
+}
+
+LoopResult CSelftestPart_FirstLayer::stateHome() {
+    return queue.enqueue_one("G28") ? LoopResult::RunNext : LoopResult::RunCurrent;
+}
+
+LoopResult CSelftestPart_FirstLayer::stateMbl() {
+    return queue.enqueue_one("G29") ? LoopResult::RunNext : LoopResult::RunCurrent;
+}
+
+LoopResult CSelftestPart_FirstLayer::statePrint() {
+    how_many_times_finished = FirstLayer::HowManyTimesFinished();
+    queue.enqueue_one_now("G26"); // firstlay
```

Comment:
faillure not hadled

Reply:
enqueue_one_now one does not fail, but can be blocking, which is wrong
will fix it